### DISCUSSION TITLE
Support composable schema with refinements

### DISCUSSION THREAD

**Discussion by JRGranell** [colinhacks/zod#955] [Q&A] [answered]
Hi, 

Thanks for the work on this library, it's great and I'm looking to bring it into our development tool chain. I had a quick question about using `refine`. I'd like to compose validation together, preferably through spreading (in or order help TS). For example:

```ts
const dateRange =  z.object({
        min: z.date(),
        max: z.date()
    })
    // Useful validation that I'd like to keep alongside the schema definition
    // so whenever I use "dateRange" this validation will always be included
    .refine(({ min, max }) => min < max);

// Compose up one schema
const params = def.object({
    id: z.string(),
    ...dateRange // or ...dateRange.shape etc.
});

// Compose up another schema
const params2 = def.object({
    name: z.string(),
    ...dateRange.shape
});
```

I understand that refine returns `ZodEffect` which[ doesn't have the shape property](https://github.com/colinhacks/zod/issues/830#issuecomment-998850309), although I don't quite understand how refine potentially alters input / output, I thought it was additional validation only?

Perhaps `.merge` offers an alternative route. However using TypeScript I was unable to use `.merge` to enforce the required keys (see [this approach](https://github.com/colinhacks/zod/issues/858#issuecomment-1016632896) to how I'm enforcing types). 

I just wondered if there were any other suggestions I could try?

Thanks in advance

**Comment by JRGranell** [comment]
Looking at the problem again, it may not make sense to spread the shape - you would still loose the refinements that were added. Also with `.merge` you can only merge objects, which means we would loose the refinements either way.

**Comment by airjp73** [comment, marked as answer]
I'm not a maintainer or anything, but I'm pretty sure you can do this like this:

```ts
const dateRange =  z
  .object({
    min: z.date(),
    max: z.date()
  })
  .refine(({ min, max }) => min < max);

// Compose up one schema
const params = dateRange.and(z.object({
    id: z.string()
}));

// Compose up another schema
const params2 = dateRange.and(z.object({
    name: z.string()
}));
```
